components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/token/handlers/grant/RefreshGrantHandler.java

```diff
@@ -282,6 +297,70 @@ private boolean validateRefreshTokenStatus(RefreshTokenValidationDataDO validati
+
+        if (!OAuthServerConfiguration.getInstance().isValidateAuthenticatedUserForRefreshGrantEnabled()) {
+            return true;
+        }
+
+        checkAccountLockStatusOfTheUser(validationBean, oAuthTokenReqMessageContext);
+
+        return true;
+    }
+
+    private void checkAccountLockStatusOfTheUser(RefreshTokenValidationDataDO validationBean,
+                                                 OAuthTokenReqMessageContext oAuthTokenReqMessageContext)
+            throws IdentityOAuth2Exception {
+
+        AuthenticatedUser authenticatedUser = validationBean.getAuthorizedUser();
+        if (authenticatedUser != null) {
+            String username = null;
+            String tenantDomain = null;
+
+            if (authenticatedUser.isFederatedUser()) {
+                try {
+                    FederatedAssociationManager federatedAssociationManager =
+                            FrameworkUtils.getFederatedAssociationManager();
+                    OAuthAppDO oAuthAppDO =
+                            (OAuthAppDO) oAuthTokenReqMessageContext.getProperty(AccessTokenIssuer.OAUTH_APP_DO);
+                    String oAuthAppTenantDomain = OAuth2Util.getTenantDomainOfOauthApp(oAuthAppDO);
+                    String associatedLocalUsername =
+                            federatedAssociationManager.getUserForFederatedAssociation(oAuthAppTenantDomain,
+                                    authenticatedUser.getFederatedIdPName(),
+                                    authenticatedUser.getAuthenticatedSubjectIdentifier());
+                    if (associatedLocalUsername != null) {
+                        username = associatedLocalUsername;
+                        tenantDomain = oAuthAppTenantDomain;
+                    }
+                } catch (FederatedAssociationManagerException | FrameworkException e) {
+                    throw new IdentityOAuth2Exception(ERROR_WHILE_GETTING_USERNAME_ASSOCIATED_WITH_IDP.getCode(),
+                            String.format(ERROR_WHILE_GETTING_USERNAME_ASSOCIATED_WITH_IDP.getMessage(),
+                                    authenticatedUser.getFederatedIdPName()), e);
+                }
+            } else {
+                username = UserCoreUtil.addDomainToName(authenticatedUser.getUserName(),
+                        authenticatedUser.getUserStoreDomain());
+                tenantDomain = authenticatedUser.getTenantDomain();
+            }
+
+            if (username != null && tenantDomain != null) {
+                AccountLockService accountLockService = OAuth2ServiceComponentHolder.getAccountLockService();
+
+                try {
+                    boolean accountLockStatus = accountLockService.isAccountLocked(username, tenantDomain);
+                    if (accountLockStatus) {
+                        throw new IdentityOAuth2Exception(UserCoreConstants.ErrorCode.USER_IS_LOCKED,
+                                String.format(ACCOUNT_LOCK_ERROR_MESSAGE, username, tenantDomain));
```

Comment:
Will it be PII to log username. Can't we throw IdentityOAuth2ClientException

Reply:
Refactor the error handling https://github.com/wso2-extensions/identity-inbound-auth-oauth/pull/2581/commits/00de5bc9dbce1a44e9159f7b32ae640fe7e34e87. Thank you for mentioning @sadilchamishka